client-device-connection-infinispan/src/main/java/org/eclipse/hono/deviceconnection/infinispan/client/HotrodCache.java

```diff
@@ -0,0 +1,279 @@
+
+import org.eclipse.hono.client.ConnectionLifecycle;
+import org.eclipse.hono.client.DisconnectListener;
+import org.eclipse.hono.client.ReconnectListener;
+import org.eclipse.hono.client.ServerErrorException;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.commons.api.BasicCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A HotrodCache.
+ *
+ * @param <K> The type of keys used by the cache.
+ * @param <V> The type of values stored in the cache.
+ */
+public final class HotrodCache<K, V> implements ConnectionLifecycle<HotrodCache<K, V>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HotrodCache.class);
+
+    private final AtomicBoolean connecting = new AtomicBoolean(false);
+    private final Vertx vertx;
+    private final RemoteCacheManager cacheManager;
+    private final String cacheName;
+    private final K connectionCheckKey;
+    private final V connectionCheckValue;
+
+    private RemoteCache<K, V> cache;
+
+    /**
+     * @param vertx The vert.x instance to run on.
+     * @param cacheManager The connection to the remote cache.
+     * @param name The name of the (remote) cache.
+     * @param connectionCheckKey The key to use for checking the connection
+     *                           to the data grid.
+     * @param connectionCheckValue The value to use for checking the connection
+     *                           to the data grid.
+     */
+    public HotrodCache(
+            final Vertx vertx,
+            final RemoteCacheManager cacheManager,
+            final String name,
+            final K connectionCheckKey,
+            final V connectionCheckValue) {
+        this.vertx = Objects.requireNonNull(vertx);
+        this.cacheManager = Objects.requireNonNull(cacheManager);
+        this.cacheName = Objects.requireNonNull(name);
+        this.connectionCheckKey = Objects.requireNonNull(connectionCheckKey);
+        this.connectionCheckValue = Objects.requireNonNull(connectionCheckValue);
+    }
+
+    public BasicCache<K, V> getCache() {
+        return cache;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Future<HotrodCache<K, V>> connect() {
+        return connectToGrid().map(ok -> this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Future<Void> isConnected() {
+        return checkForCacheAvailability().mapEmpty();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void disconnect() {
+        disconnect(r -> {});
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void disconnect(final Handler<AsyncResult<Void>> completionHandler) {
+
+        vertx.executeBlocking(r -> {
```

Comment:
Why not use RemoteCache or AdvanceCache interfaces and then use "async" variant of the calls instead of handling this manually?

Reply:
I originally had that but the code is not much different because you still need to adapt the standard Java `CompletableFuture` to vert.x `Promise` ...

Reply:
OK, makes sense.